### PR TITLE
Extra data in block structure api and indexer

### PR DESCRIPTION
**Pull request by miiu96**
- Upgrade the indexer version ( this new version will save extra fields in the `block` structure)
- Extend also the API block structure

### CODE REVIEW

**Review by bogdan-rosianu — APPROVED** [review]

**Review by popenta — APPROVED** [review]
@@                            Log scanner                                  @@
##                            indexer-extra-data-block-structure           ##
================================================================================
- Known Warnings              18
- New Warnings                0
- Known Errors                0
- New Errors                  3
- Panics                      26
================================================================================

**Review by bogdan-rosianu — APPROVED** [review]

File: node/external/blockAPI/metaBlock.go [hunk 1]
```diff
@@ -152,25 +152,74 @@ func (mbp *metaAPIBlockProcessor) convertMetaBlockBytesToAPIBlock(hash []byte, b
...
+		PrevEpochStartRound:              epochStartEconomics.PrevEpochStartRound,
+		PrevEpochStartHash:               hex.EncodeToString(epochStartEconomics.PrevEpochStartHash),
+	}
+
+	if len(metaBlock.EpochStart.LastFinalizedHeaders) == 0 {
+		return
+	}
+
+	epochStartShardsData := metaBlock.EpochStart.LastFinalizedHeaders
+	apiBlock.EpochStartShardsData = make([]*api.EpochStartShardData, 0, len(metaBlock.EpochStart.LastFinalizedHeaders))
```
Comment (bogdan-rosianu) [inline node/external/blockAPI/metaBlock.go:188]: could have used `len(epochStartShardsData)`

Comment (miiu96) [reply node/external/blockAPI/metaBlock.go]: done.

File: node/external/blockAPI/baseBlock.go [hunk 2]
```diff
@@ -324,3 +325,27 @@ func extractExecutedTxHashes(mbTxHashes [][]byte, firstProcessed, lastProcessed
 
 	return mbTxHashes[firstProcessed : lastProcessed+1]
 }
+
+func addScheduledInfoInBlock(header data.HeaderHandler, apiBlock *api.Block) {
+	additionalData := header.GetAdditionalData()
+	if additionalData == nil {
```
Comment (iulianpascalau) [inline node/external/blockAPI/baseBlock.go:331]: `check.IfNil(additionalData)` as `header.GetAdditionalData()` will return an interface

File: node/external/blockAPI/baseBlock.go [hunk 3]
```diff
@@ -324,3 +325,27 @@ func extractExecutedTxHashes(mbTxHashes [][]byte, firstProcessed, lastProcessed
 
 	return mbTxHashes[firstProcessed : lastProcessed+1]
 }
+
+func addScheduledInfoInBlock(header data.HeaderHandler, apiBlock *api.Block) {
```
Comment (iulianpascalau) [inline node/external/blockAPI/baseBlock.go:329]: unit tests

File: node/external/blockAPI/baseBlock.go [hunk 4]
```diff
@@ -324,3 +325,27 @@ func extractExecutedTxHashes(mbTxHashes [][]byte, firstProcessed, lastProcessed
...
+		ScheduledRootHash:        hex.EncodeToString(additionalData.GetScheduledRootHash()),
+		ScheduledAccumulatedFees: bigIntToStr(additionalData.GetScheduledAccumulatedFees()),
+		ScheduledDeveloperFees:   bigIntToStr(additionalData.GetScheduledDeveloperFees()),
+		ScheduledGasProvided:     additionalData.GetScheduledGasProvided(),
+		ScheduledGasPenalized:    additionalData.GetScheduledGasPenalized(),
+		ScheduledGasRefunded:     additionalData.GetScheduledGasRefunded(),
+	}
+}
+
+func bigIntToStr(value *big.Int) string {
```
Comment (iulianpascalau) [inline node/external/blockAPI/baseBlock.go:345]: unit tests

File: node/external/blockAPI/metaBlock.go [hunk 5]
```diff
@@ -152,25 +152,74 @@ func (mbp *metaAPIBlockProcessor) convertMetaBlockBytesToAPIBlock(hash []byte, b
...
-			RewardsPerBlock:                  epochStartEconomics.RewardsPerBlock.String(),
-			RewardsForProtocolSustainability: epochStartEconomics.RewardsForProtocolSustainability.String(),
-			NodePrice:                        epochStartEconomics.NodePrice.String(),
-			PrevEpochStartRound:              epochStartEconomics.PrevEpochStartRound,
-			PrevEpochStartHash:               hex.EncodeToString(epochStartEconomics.PrevEpochStartHash),
-		}
+	return apiMetaBlock, nil
+}
+
+func addStartOfEpochInfoInBlock(metaBlock *block.MetaBlock, apiBlock *api.Block) {
```
Comment (iulianpascalau) [inline node/external/blockAPI/metaBlock.go:165]: :+1:

File: node/external/blockAPI/metaBlock.go [hunk 6]
```diff
@@ -152,25 +152,74 @@ func (mbp *metaAPIBlockProcessor) convertMetaBlockBytesToAPIBlock(hash []byte, b
...
+		Nonce:                 epochStartShardData.Nonce,
+		HeaderHash:            hex.EncodeToString(epochStartShardData.HeaderHash),
+		RootHash:              hex.EncodeToString(epochStartShardData.RootHash),
+		ScheduledRootHash:     hex.EncodeToString(epochStartShardData.ScheduledRootHash),
+		FirstPendingMetaBlock: hex.EncodeToString(epochStartShardData.FirstPendingMetaBlock),
+		LastFinishedMetaBlock: hex.EncodeToString(epochStartShardData.LastFinishedMetaBlock),
+	}
+
+	if len(epochStartShardData.PendingMiniBlockHeaders) == 0 {
+		apiBlock.EpochStartShardsData = append(apiBlock.EpochStartShardsData, shardData)
```
Comment (iulianpascalau) [inline node/external/blockAPI/metaBlock.go:208]: move this outside the if branch and remove L222

File: node/external/blockAPI/metaBlock_test.go [hunk 7]
```diff
@@ -418,7 +418,26 @@ func TestMetaAPIBlockProcessor_GetBlockByRound_GetBlockByNonce_EpochStartBlock(t
 			},
 		},
 		EpochStart: block.EpochStart{
-			LastFinalizedHeaders: []block.EpochStartShardData{{}},
+			LastFinalizedHeaders: []block.EpochStartShardData{{
```
Comment (iulianpascalau) [inline node/external/blockAPI/metaBlock_test.go:421]: Please do:
* add one more finalized header
* add one more miniblock header in each finalized header

**Review by iulianpascalau — CHANGES_REQUESTED** [review]

File: node/external/blockAPI/baseBlock.go (hunk 2 above)
Comment (miiu96) [reply node/external/blockAPI/baseBlock.go:331]: done.

File: node/external/blockAPI/baseBlock.go (hunk 3 above)
Comment (miiu96) [reply node/external/blockAPI/baseBlock.go:329]: added

File: node/external/blockAPI/baseBlock.go (hunk 4 above)
Comment (miiu96) [reply node/external/blockAPI/baseBlock.go:345]: added

File: node/external/blockAPI/metaBlock.go (hunk 6 above)
Comment (miiu96) [reply node/external/blockAPI/metaBlock.go:208]: done.

File: node/external/blockAPI/metaBlock_test.go (hunk 7 above)
Comment (miiu96) [reply node/external/blockAPI/metaBlock_test.go]: done.

**Review by iulianpascalau — APPROVED** [review]

**Review by bogdan-rosianu — APPROVED** [review]

**Review by popenta — APPROVED** [review]
@@                            Log scanner                                  @@
##                            indexer-extra-data-block-structure           ##
================================================================================
- Known Warnings              14
- New Warnings                0
- Known Errors                0
- New Errors                  0
- Panics                      0
================================================================================

**Pull request merged**